### PR TITLE
Added support for local binding. Closes #122

### LINKED ISSUE #122: [FR] Bind source address.
I need to call [bind()](https://www.boost.org/doc/libs/1_47_0/doc/html/boost_asio/reference/basic_stream_socket/bind.html) on the underlying scoket to send requests from a specific NIC.

### PR DESCRIPTION
**Pull request by jgaa**

### CODE REVIEW

File: src/RequestImpl.cpp [hunk 1]
```diff
@@ -292,10 +291,92 @@ class RequestImpl : public Request {
...
+        }
+
+        return p;
+    }
+
     Connection::ptr_t Connect(Context& ctx) {
 
         static const auto timer_name = "Connect"s;
 
+        auto prot_filter = GetBindProtocols(properties_->bindToLocalAddress, ctx);
```
Comment (gou4shi1) [inline src/RequestImpl.cpp:378]: Only call `GetBindProtocols()`  under `if (!connection->GetSocket().IsOpen())` to save cost?

Comment (jgaa) [reply src/RequestImpl.cpp:378]: It's the work-flow for new connections. All connection object instances are maintained by a connection-pool, and this block opens the connection if it is new.

Comment (gou4shi1) [reply src/RequestImpl.cpp:378]: Yes, but `Connect()` is called for every `SendRequest()`, so `GetBindProtocols()` is called for every time also. So I suggest to only call `GetBindProtocols()` if `!connection->GetSocket().IsOpen()`, i.e. move line 378 to line 411

Comment (gou4shi1) [reply src/RequestImpl.cpp:378]: I thought we could move this line to line 411 to save cost if the connection is already open, but there is a loop in line 396.

**Pull request merged**